backend/app/notify/email.go

```diff
@@ -0,0 +1,380 @@
+	FlushDuration        time.Duration // maximum time after which email will me sent, 30s by default
+}
+
+// Email implements notify.Destination for email
+type Email struct {
+	EmailParams
+	smtpClient // embedded into Email for test purposes: not set in production code, used in autoFlush initialisation
+
+	msgTmpl    *template.Template // parsed request message template
+	verifyTmpl *template.Template // parsed verification message template
+	submit     chan emailMessage  // unbuffered channel for email sending
+	once       sync.Once
+}
+
+// smtpClient interface defines subset of net/smtp used by email client
+type smtpClient interface {
+	Mail(string) error
+	Auth(smtp.Auth) error
+	Rcpt(string) error
+	Data() (io.WriteCloser, error)
+	Quit() error
+	Close() error
+}
+
+type emailMessage struct {
+	message string
+	to      string
+}
+
+// msgTmplData store data for message from request template execution
+type msgTmplData struct {
+	From      string
+	To        string
+	Orig      string
+	Link      string
+	PostTitle string
+}
+
+// verifyTmplData store data for verification message template execution
+type verifyTmplData struct {
+	User  string
+	Email string
+	Token string
+	Site  string
+}
+
+const defaultEmailTimeout = 10 * time.Second
+const defaultFlushDuration = time.Second * 30
+const defaultEmailTemplate = `{{.From}}{{if .To}} → {{.To}}{{end}}
+
+{{.Orig}}
+
+↦ <a href="{{.Link}}">{{if .PostTitle}}{{.PostTitle}}{{else}}original comment{{end}}</a>
+`
+
+const defaultEmailVerificationTemplate = `Confirmation for {{.User}} {{.Email}}, site {{.Site}}
+
+Token: {{.Token}}
+`
+
+//NewEmail makes new Email object, returns it even in case of problems
+// (e.MsgTemplate parsing error or error while testing smtp connection by credentials provided in params)
+func NewEmail(params EmailParams) (*Email, error) {
+	var err error
+	res := Email{EmailParams: params}
+	if res.TimeOut <= 0 {
+		res.TimeOut = defaultEmailTimeout
+	}
+	if res.FlushDuration <= 0 {
+		res.FlushDuration = defaultFlushDuration
+	}
+	if res.BufferSize <= 0 {
+		res.BufferSize = 1
+	}
+	if res.MsgTemplate == "" {
+		res.MsgTemplate = defaultEmailTemplate
+	}
+	if res.VerificationTemplate == "" {
+		res.VerificationTemplate = defaultEmailVerificationTemplate
+	}
+	// unbuffered send channel for sending messages to autoFlush goroutine
+	res.submit = make(chan emailMessage)
+
+	log.Printf("[DEBUG] create new email notifier for server %s with user %s, timeout=%s",
+		res.Host, res.Username, res.TimeOut)
+
+	// initialise templates
+	res.msgTmpl, err = template.New("messageFromRequest").Parse(res.MsgTemplate)
+	if err != nil {
+		return &res, errors.Wrapf(err, "can't parse message template")
+	}
+	res.verifyTmpl, err = template.New("messageFromRequest").Parse(res.VerificationTemplate)
+	if err != nil {
+		return &res, errors.Wrapf(err, "can't parse verification template")
+	}
+
+	// establish test connection
+	testSMTPClient, err := res.client()
```

Comment:
I would rather have it here to have a timely error message in case of a problem and the distinction of temporary email server connection problems from permanent ones (which is harder/impossible with lazy call).
I'm willing to pay the cost of test connection establishment on startup for not requiring the user to send an email to discover that he messed up with the configuration. What is your opinion on this?